sdks/go/pkg/beam/runners/direct/direct.go

```diff
@@ -241,10 +231,31 @@ func (b *builder) makeLink(id linkID) (exec.Node, error) {
 		return b.links[id], nil
 
 	case graph.Combine:
-		u = &exec.Combine{UID: b.idgen.New(), Edge: edge, Out: out[0]}
+		isPerKey := typex.IsWCoGBK(edge.Input[0].From.Type())
+		usesKey := typex.IsWKV(edge.Input[0].Type)
+
+		u = &exec.Combine{UID: b.idgen.New(), Fn: edge.CombineFn, IsPerKey: isPerKey, UsesKey: usesKey, Out: out[0]}
+
+	case graph.CoGBK:
+		u = &CoGBK{UID: b.idgen.New(), Edge: edge, Out: out[0]}
+		b.units = append(b.units, u)
 
-	case graph.GBK:
-		u = &GBK{UID: b.idgen.New(), Edge: edge, Out: out[0]}
+		// CoGBK needs injection of each incoming index. If > 1 incoming,
+		// insert Flatten as well.
```

Comment:
<!--thread_id:cc_166114558_t; commit:977a4f133a10d957b90566426581193d93041321; resolved:1-->
<!--section:context-quote-->
> **wcn3** wrote:
> nit: insert Flatten first. 'as well' implies a following relationship which isn't accurate here.

<!--section:body-->
Flatten does follow the Injects: the plan is built bottom-up.

Reply:
Got it. I was ordering things the wrong way.